### PR TITLE
Enhancement proposal to add minReplicas per VPA Object (see #4560)

### PR DESCRIPTION
**Pull request by kgolab**
#### Which component this PR applies to?

vertical-pod-autoscaler (VPA API)

#### What type of PR is this?

/kind documentation
/kind feature
/kind api-change

#### What this PR does / why we need it:

To allow for API review before merging (or rejecting) #4560.

### CODE REVIEW

**Comment by kgolab** [comment]
@mwielgus , you asked for the KEP to accompany my PR #4560. Please have a look if this is what you had in mind. Thanks!

**Comment by kgolab** [comment]
As per yesterday's SIG meeting: @jbartosik , @wangchen615, you might be interested in looking into this.

File: vertical-pod-autoscaler/enhancements/4566-min-replicas/README.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+The proposal is to add `minReplicas` field under VPA `spec.updatePolicy` and
+alter VPA Updater behaviour so it respects the new field:
+- if the field is set, use its value instead of the global `--min-replicas`
+  flag, only for this particular VPA object,
+- if the field is not set, use the global `--min-replicas` flag, keeping
+  backward compatibility.
+
+The field is ignored by other VPA components (Recommender, Admission
+ontroller).
```
Comment (jbartosik) [inline vertical-pod-autoscaler/enhancements/4566-min-replicas/README.md:59]: Nit: dropped first letter

File: vertical-pod-autoscaler/enhancements/4566-min-replicas/README.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+An alternative suggested in
+[issue #3986](https://github.com/kubernetes/autoscaler/issues/3986) was to reuse
+Cluster Autoscaler annotation `cluster-autoscaler.kubernetes.io/safe-to-evict`.
+
+While it's not clear if there are use cases when one would like to allow VPA or
+CA to evict Pods but disallow it for the other controller, coupling them this
+way seems dangerous.
+
+Of course VPA could've introduced own annotation to express the same but it
```
Comment (jbartosik) [inline vertical-pod-autoscaler/enhancements/4566-min-replicas/README.md:103]: Nit: I'd simplify grammar a bit: "VPA could introduce it's own annotation to express ..."

File: vertical-pod-autoscaler/enhancements/4566-min-replicas/README.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+## Alternatives
+
+### Existing Behaviour
+
+The only existing alternative to achieve the behaviour allowed by the proposed
+change is to set global `--min-replicas` flag to `1` and Pod Disruption Budget
+to protect single-replica workloads from being disrupted, if needed.
+
+While this would end up in a very similar (if not identical) cluster behaviour,
```
Comment (jbartosik) [inline vertical-pod-autoscaler/enhancements/4566-min-replicas/README.md:87]: I think cluster behavior is different.

With this KEP we can protect 1 pod controllers from eviction by default (and can override that per-VPA).
With the alternative described in this section we evict 1-pod controllers by default (and can override that per-controller).

I'd rephrase to something like:

While this allows the same configuration for each individual controller it changes the default behavior to an unsafe one. As a result it requires cluster-wide coordination...

**Review by jbartosik — APPROVED** [review]
Generally looks good to me, I left some small comments about phrasing.

File: vertical-pod-autoscaler/enhancements/4566-min-replicas/README.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+   - [Existing Behaviour](#existing-behaviour)
+   - [Reuse Cluster Autoscaler Annotations](#reuse-cluster-autoscaler-annotations)
+<!-- /toc -->
+
+## Summary
+
+The default behaviour of VPA Updater is to allow Pod eviction only if there are
+at least 2 live replicas, in order to avoid temporaty total unavailability of a
+workload under VPA in Auto mode. This can be changed globally with
+`--min-replicas` flag. However such a change might be deemed dangerous in a
```
Comment (wangchen615) [inline vertical-pod-autoscaler/enhancements/4566-min-replicas/README.md:22]: However such => However, such ...

File: vertical-pod-autoscaler/enhancements/4566-min-replicas/README.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+The motivation behind the change is due to VPA users needing the configuration
+option in managed environments to allow for eviction of a single replica, see
+[issue #3986](https://github.com/kubernetes/autoscaler/issues/3986) or
+[issue #1828](https://github.com/kubernetes/autoscaler/issues/1828).
+
+### Goals
+
+- Main: allow workload owner to use VPA in Auto/Recreate mode on a single-replica
+  workloads,
+- Secondary: allow workload owner to specify higher number of replicas required
```
Comment (wangchen615) [inline vertical-pod-autoscaler/enhancements/4566-min-replicas/README.md:41]: specify higher => specify a higher number ....

**Review by wangchen615 — COMMENTED** [review]
Just some minor issues. I like the idea, thanks.

File: vertical-pod-autoscaler/enhancements/4566-min-replicas/README.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+version of VPA API, avoiding the hassle of introducing a new API version.
+
+## Design Details
+
+Suggested implementation is present in [PR
+4560](https://github.com/kubernetes/autoscaler/pull/4560).
+
+### Test Plan
+
+Add automated tests that update VPA objects, altering the value of
```
[the remark this answers is not in the document]
Comment (kgolab) [reply vertical-pod-autoscaler/enhancements/4566-min-replicas/README.md]: Actually they're already there in PR #4560, please have a look at e2e/v1/actuation.go. The testEvictsSingletonPod verifies exactly the behaviour described here: no eviction w/o minReplicas being set and eviction happening after minReplicas is set to 1.

Added 'E2E' in the KEP to make it clear.

File: vertical-pod-autoscaler/enhancements/4566-min-replicas/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (kgolab) [reply vertical-pod-autoscaler/enhancements/4566-min-replicas/README.md]: Fixed.

File: vertical-pod-autoscaler/enhancements/4566-min-replicas/README.md (hunk 5 above)
Comment (kgolab) [reply vertical-pod-autoscaler/enhancements/4566-min-replicas/README.md]: Thanks!

File: vertical-pod-autoscaler/enhancements/4566-min-replicas/README.md (hunk 4 above)
Comment (kgolab) [reply vertical-pod-autoscaler/enhancements/4566-min-replicas/README.md]: Ach, commas...

File: vertical-pod-autoscaler/enhancements/4566-min-replicas/README.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+- [Implementation History](#implementation-history)
+- [Alternatives](#alternatives)
+   - [Existing Behaviour](#existing-behaviour)
+   - [Reuse Cluster Autoscaler Annotations](#reuse-cluster-autoscaler-annotations)
+<!-- /toc -->
+
+## Summary
+
+The default behaviour of VPA Updater is to allow Pod eviction only if there are
+at least 2 live replicas, in order to avoid temporaty total unavailability of a
```
[the remark this answers is not in the document]
Comment (kgolab) [reply vertical-pod-autoscaler/enhancements/4566-min-replicas/README.md]: Thanks!

File: vertical-pod-autoscaler/enhancements/4566-min-replicas/README.md (hunk 2 above)
Comment (kgolab) [reply vertical-pod-autoscaler/enhancements/4566-min-replicas/README.md]: Done.

File: vertical-pod-autoscaler/enhancements/4566-min-replicas/README.md (hunk 1 above)
Comment (kgolab) [reply vertical-pod-autoscaler/enhancements/4566-min-replicas/README.md]: Thanks.

**Comment by kgolab** [comment]
@jbartosik , @wangchen615 . hopefully all your comments are addressed. Thanks for the review :-)

**Review by jbartosik — APPROVED** [review]
Looks good to me.

I'm not giving the label yet because that would merge the PR and I want to give others some time to take a look. (feel free to ping me if you think it's time to merge but I didn't give the label yet)

/retest

**Comment by jbartosik** [comment]
@wangchen615 please let me know you want to take another look (a comment telling me I should wait more is enough). If this PR doesn't get any more comments I'll give it LGTM and it'll merge on Monday (2022-01-17) during my work hours.

I'd like to merge this, then #4560 then do a release #4404. I want to make a release next week. Or if we need more discussion on this I'd like to know that by Monday so that I can go ahead and release without waiting for this.

**Comment by wangchen615** [comment]
> @wangchen615 please let me know you want to take another look (a comment telling me I should wait more is enough). If this PR doesn't get any more comments I'll give it LGTM and it'll merge on Monday (2022-01-17) during my work hours.
> 
> I'd like to merge this, then #4560 then do a release #4404. I want to make a release next week. Or if we need more discussion on this I'd like to know that by Monday so that I can go ahead and release without waiting for this.

@jbartosik I already went through it. It looked good to me. No more comments.

**Review by wangchen615 — APPROVED** [review]

**Review by jbartosik — COMMENTED** [review]
Thanks :)

/lgtm
